File: sistem-kamera-pengawas/src/app/pages/people/people.ts
import { Component, DestroyRef, ElementRef, inject, OnInit } from '@angular/core';
import { NgClass } from '@angular/common';
import { HttpClient } from '@angular/common/http';
import { ActivatedRoute, Router } from '@angular/router';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { Subject } from 'rxjs';
import { debounceTime, distinctUntilChanged } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import { AuthUrlPipe } from '../../pipes/auth-url.pipe';

const API = environment.apiBaseUrl;

// ── Feed mentah, dari GET /people/feed (level deteksi) — section DETEKSI ────

interface FeedItem {
  detection_id:  number;
  person_id:     number | null;
  person_label:  string | null;
  person_name:   string | null;
  is_known:      boolean;
  camera_id:       string;
  camera_name:     string | null;
  camera_location: string | null;
  timestamp:       string;
  thumbnail_url:   string | null;
  tracklet_id:     number | null;
}

interface FeedResponse {
  items: FeedItem[];
  total: number;
  page:  number;
  pages: number;
  limit: number;
}

// ── Card ORANG — dari GET /people/persons (level orang, paginasi sendiri,
// independen dari section DETEKSI). "Beri Nama" ada di sini, bukan di card
// Deteksi.
interface PersonCard {
  person_id:     number;
  person_name:   string | null;
  person_label:  string | null;
  is_known:        boolean;
  thumbnail_url:   string | null;
  camera_name:     string | null;
  camera_location: string | null;
  camera_id:       string;
  timestamp:       string;
  tracklet_id:     number | null;
}

interface Camera {
  id:         number;
  camera_id:  string | null;
  name:       string;
  location:   string | null;
  group_name: string | null;
}

interface CameraGroup {
  name:    string;
  cameras: Camera[];
}

interface NameSuggestion {
  person_id:     number;
  name:          string;
  jabatan:       string | null;
  similarity:    number;
  thumbnail_url: string | null;
}

// Sama persis dengan CLIP_COLORS di ai-service/app/par/par_service.py — value
// yang dikirim ke backend harus cocok string yang disimpan PAR, apa adanya.
// swatch = warna CSS approksimasi buat lingkaran pratinjau.
const ATTR_COLORS: { value: string; label: string; swatch: string }[] = [
  { value: 'black',  label: 'Hitam',      swatch: '#1a1a1a' },
  { value: 'white',  label: 'Putih',      swatch: '#f5f5f5' },
  { value: 'gray',   label: 'Abu-abu',    swatch: '#8a8a8a' },
  { value: 'red',    label: 'Merah',      swatch: '#dc2626' },
  { value: 'green',  label: 'Hijau',      swatch: '#16a34a' },
  { value: 'blue',   label: 'Biru',       swatch: '#2563eb' },
  { value: 'brown',  label: 'Coklat',     swatch: '#78350f' },
  { value: 'yellow', label: 'Kuning',     swatch: '#eab308' },
  { value: 'purple', label: 'Ungu',       swatch: '#9333ea' },
  { value: 'pink',   label: 'Merah muda', swatch: '#ec4899' },
];

// Aksesoris — tiap entri bisa mewakili >1 atribut PAR mentah yang di-OR
// (mis. tas = gabungan 3 jenis tas). Independen satu sama lain (AND).
const ACCESSORIES: { id: string; label: string; attrNames: string[] }[] = [
  { id: 'bag',        label: 'Tas',      attrNames: ['attach backpack', 'attach shoulder bag', 'attach hand bag'] },
  { id: 'hat',        label: 'Topi',     attrNames: ['head hat'] },
  { id: 'sunglasses', label: 'Kacamata', attrNames: ['head glasses'] },
];

// State terakhir sebelum user membuka Person Investigation — dipulihkan saat
// kembali ke /people. Hilang saat full reload (memang: reload = mulai bersih).
// ponytail: module var, cukup satu file — bukan RouteReuseStrategy.
interface PeopleSnapshot {
  activeTab: 'deteksi' | 'orang';
  searchQuery: string;
  selectedDate: string;
  selectedCameras: string[];
  upperColors: string[];
  lowerColors: string[];
  gender: 'male' | 'female' | null;
  selectedAccessories: string[];
  similarTo: number | null;
  page: number;
  orangPage: number;
  scrollY: number;
}
let savedState: PeopleSnapshot | null = null;

@Component({
  selector: 'app-people',
  standalone: true,
  imports: [NgClass, AuthUrlPipe],
  templateUrl: './people.html',
  styleUrl: './people.css',
})
export class People implements OnInit {
  private http       = inject(HttpClient);
  private router     = inject(Router);
  private route      = inject(ActivatedRoute);
  private destroyRef = inject(DestroyRef);
  private host       = inject(ElementRef);

  // Kontainer scroll adalah .content (app.css), bukan window.
  private get _scroller(): Element | null {
    return (this.host.nativeElement as HTMLElement).closest('.content');
  }

  // Tab aktif di bawah search/filter — hanya satu section tampil sekaligus.
  activeTab: 'deteksi' | 'orang' = 'deteksi';

  // ── Search & filter state ─────────────────────────────────────────────────
  searchQuery = '';
  selectedDate = this._todayISO();

  cameras: Camera[] = [];
  cameraGroups: CameraGroup[] = [];
  selectedCameras: string[] = [];
  showCameraPanel = false;

  // Filter atribut (Fase 3, dari PAR) — kosong/null = tidak difilter.
  upperColors: string[] = [];
  lowerColors: string[] = [];
  gender: 'male' | 'female' | null = null;
  selectedAccessories: string[] = [];
  showAttrPanel = false;
  readonly attrColors  = ATTR_COLORS;
  readonly accessories = ACCESSORIES;

  // Appearance search ("Cari Serupa") — dipicu dari halaman investigasi.
  similarTo: number | null = null;

  // ── Section DETEKSI — raw, paginated, urut waktu ────────────────────────────
  deteksi: FeedItem[] = [];
  total = 0;
  page  = 1;
  pages = 1;

  // ── Section ORANG — GET /people/persons, paginasi SENDIRI (independen dari
  // DETEKSI) — filter yang dipakai sama, tapi halaman/total dihitung per orang.
  orang:      PersonCard[] = [];
  orangTotal = 0;
  orangPage  = 1;
  orangPages = 1;

  private readonly limit = 24;   // 6 kolom × 4 baris per halaman (lihat .orang-grid/.deteksi-grid)

  // ── "Beri nama" inline form — di card ORANG, bukan per-deteksi ─────────────
  namingPersonId: number | null = null;
  nameInput    = '';
  jabatanInput = '';
  nameSuggestions: NameSuggestion[] = [];

  private search$ = new Subject<void>();

  // scrollTop yang menunggu dipulihkan setelah data render (lihat _loadFeed/_loadPersons)
  private _pendingScroll: number | null = null;

  ngOnInit(): void {
    this._loadCameras();

    const restored = savedState;
    savedState = null;
    if (restored) {
      const { scrollY, ...state } = restored;
      Object.assign(this, state);
    }

    const qp = this.route.snapshot.queryParamMap.get('similar_to');
    if (qp) this.similarTo = Number(qp);

    const q = this.route.snapshot.queryParamMap.get('q');
    if (q) this.searchQuery = q;

    this.search$.pipe(
      debounceTime(300),
      distinctUntilChanged(),
      takeUntilDestroyed(this.destroyRef),
    ).subscribe(() => this._reload());

    if (restored) {
      // pulihkan tanpa reset page/tab; scroll dipulihkan setelah data render
      this._pendingScroll = restored.scrollY;
      this._loadFeed();
      this._loadPersons();
    } else {
      this._reload();
    }

    // Auto-refresh: deteksi baru dari AI service masuk terus. Poll halaman
    // saat ini (bukan _reload — biar page & scroll gak ke-reset), cuma kalau
    // tab aktif & lagi lihat hari ini. Selain itu percuma, data lama gak berubah.
    const poll = setInterval(() => {
      if (document.hidden || this.selectedDate !== this._todayISO()) return;
      this._loadFeed();
      this._loadPersons();
    }, 30_000);
    this.destroyRef.onDestroy(() => clearInterval(poll));
  }

  private _loadCameras(): void {
    this.http.get<Camera[]>(`${API}/cameras`).subscribe({
      next: cams => {
        this.cameras = cams;
        const byGroup = new Map<string, Camera[]>();
        for (const cam of cams) {
          const key = cam.group_name || 'Tanpa Grup';
          if (!byGroup.has(key)) byGroup.set(key, []);
          byGroup.get(key)!.push(cam);
        }
        this.cameraGroups = Array.from(byGroup.entries()).map(([name, cameras]) => ({ name, cameras }));
      },
      error: err => console.error('[people] cameras error:', err),
    });
  }

  // ── Search bar ────────────────────────────────────────────────────────────

  onSearchInput(event: Event): void {
    this.searchQuery = (event.target as HTMLInputElement).value;
    this.search$.next();
  }

  onSearchKeydown(event: KeyboardEvent): void {
    if (event.key === 'Enter') this._reload();
  }

  onDateChange(event: Event): void {
    this.selectedDate = (event.target as HTMLInputElement).value || this._todayISO();
    this._reload();
  }

  // ── Kamera: checklist multi-select berkelompok ───────────────────────────

  toggleCamera(camId: string): void {
    this.selectedCameras = this.selectedCameras.includes(camId)
      ? this.selectedCameras.filter(c => c !== camId)
      : [...this.selectedCameras, camId];
    this._reload();
  }

  isCameraSelected(camId: string): boolean {
    return this.selectedCameras.includes(camId);
  }

  clearCameras(): void {
    this.selectedCameras = [];
    this._reload();
  }

  // ── Atribut: warna (multi/OR), gender (single), aksesoris (independen/AND) ──

  toggleColor(kind: 'upper' | 'lower', value: string): void {
    const arr  = kind === 'upper' ? this.upperColors : this.lowerColors;
    const next = arr.includes(value) ? arr.filter(v => v !== value) : [...arr, value];
    if (kind === 'upper') this.upperColors = next; else this.lowerColors = next;
    this._reload();
  }

  isColorSelected(kind: 'upper' | 'lower', value: string): boolean {
    return (kind === 'upper' ? this.upperColors : this.lowerColors).includes(value);
  }

  setGender(value: 'male' | 'female'): void {
    this.gender = this.gender === value ? null : value;
    this._reload();
  }

  toggleAccessory(id: string): void {
    this.selectedAccessories = this.selectedAccessories.includes(id)
      ? this.selectedAccessories.filter(a => a !== id)
      : [...this.selectedAccessories, id];
    this._reload();
  }

  isAccessorySelected(id: string): boolean {
    return this.selectedAccessories.includes(id);
  }

  get hasAttrFilter(): boolean {
    return this.upperColors.length > 0 || this.lowerColors.length > 0
      || this.gender !== null || this.selectedAccessories.length > 0;
  }

  get activeAttrFilterCount(): number {
    return this.upperColors.length + this.lowerColors.length
      + (this.gender !== null ? 1 : 0) + this.selectedAccessories.length;
  }

  clearAttrFilters(): void {
    this.upperColors = [];
    this.lowerColors = [];
    this.gender = null;
    this.selectedAccessories = [];
    this._reload();
  }

  clearSimilarTo(): void {
    this.similarTo = null;
    this.router.navigate([], { relativeTo: this.route, queryParams: {} });
    this._reload();
  }

  // ── Muat data ─────────────────────────────────────────────────────────────
  // DETEKSI (/people/feed) dan ORANG (/people/persons) pakai filter yang
  // sama tapi paginasi independen — masing-masing punya page/pages sendiri.

  private _reload(): void {
    this.page = 1;
    this.orangPage = 1;
    this._loadFeed();
    this._loadPersons();
  }

  private _filterParams(page: number): URLSearchParams {
    const params = new URLSearchParams({
      page:  String(page),
      limit: String(this.limit),
      from:  this.selectedDate,
      to:    this.selectedDate,
    });
    if (this.searchQuery)            params.set('q', this.searchQuery);
    if (this.selectedCameras.length) params.set('camera_id', this.selectedCameras.join(','));
    if (this.upperColors.length)     params.set('upper_color', this.upperColors.join(','));
    if (this.lowerColors.length)     params.set('lower_color', this.lowerColors.join(','));
    if (this.gender)                 params.set('gender', this.gender);
    if (this.similarTo != null)      params.set('similar_to', String(this.similarTo));
    for (const acc of this.accessories) {
      if (this.selectedAccessories.includes(acc.id)) params.append('attrs', acc.attrNames.join(','));
    }
    return params;
  }

  private _loadFeed(): void {
    const params = this._filterParams(this.page);
    this.http.get<FeedResponse>(`${API}/people/feed?${params}`).subscribe({
      next: res => {
        this.deteksi = res.items;
        this.total   = res.total;
        this.pages   = res.pages;
        this._restoreScroll();
      },
      error: err => console.error('[people] feed error:', err),
    });
  }

  private _loadPersons(): void {
    const params = this._filterParams(this.orangPage);
    this.http.get<FeedResponse>(`${API}/people/persons?${params}`).subscribe({
      next: res => {
        this.orang      = res.items.filter((i): i is FeedItem & { person_id: number } => i.person_id != null);
        this.orangTotal = res.total;
        this.orangPages = res.pages;
        this._restoreScroll();
      },
      error: err => console.error('[people] persons error:', err),
    });
  }

  private _restoreScroll(): void {
    if (this._pendingScroll == null) return;
    const y = this._pendingScroll;
    this._pendingScroll = null;
    setTimeout(() => this._scroller?.scrollTo(0, y), 0);
  }

  goToPage(p: number): void {
    if (p < 1 || p > this.pages) return;
    this.page = p;
    this._loadFeed();
  }

  goToOrangPage(p: number): void {
    if (p < 1 || p > this.orangPages) return;
    this.orangPage = p;
    this._loadPersons();
  }

  // Windowed: selalu tampilkan halaman 1 & terakhir, plus tetangga dekat
  // halaman aktif — jangan render semua nomor kalau total halaman besar.
  private _windowedPages(cur: number, total: number): (number | '...')[] {
    if (total <= 7) return Array.from({ length: total }, (_, i) => i + 1);

    const nums: (number | '...')[] = [1];
    if (cur > 3) nums.push('...');
    for (let p = Math.max(2, cur - 1); p <= Math.min(total - 1, cur + 1); p++) nums.push(p);
    if (cur < total - 2) nums.push('...');
    nums.push(total);
    return nums;
  }

  get pageNumbers(): (number | '...')[] {
    return this._windowedPages(this.page, this.pages);
  }

  get orangPageNumbers(): (number | '...')[] {
    return this._windowedPages(this.orangPage, this.orangPages);
  }

  // ── Navigasi ke Person Investigation ─────────────────────────────────────────

  openPerson(personId: number): void {
    if (this.namingPersonId === personId) return; // form nama sedang terbuka, jangan navigasi
    savedState = {
      activeTab: this.activeTab,
      searchQuery: this.searchQuery,
      selectedDate: this.selectedDate,
      selectedCameras: this.selectedCameras,
      upperColors: this.upperColors,
      lowerColors: this.lowerColors,
      gender: this.gender,
      selectedAccessories: this.selectedAccessories,
      similarTo: this.similarTo,
      page: this.page,
      orangPage: this.orangPage,
      scrollY: this._scroller?.scrollTop ?? 0,
    };
    this.router.navigate(['/people', personId]);
  }

  // ── "Beri nama" inline, di card ORANG ────────────────────────────────────────

  startNaming(card: PersonCard, event: Event): void {
    event.stopPropagation();
    this.namingPersonId = card.person_id;
    this.nameInput    = '';
    this.jabatanInput = '';
    this.nameSuggestions = [];
    this.http.get<NameSuggestion[]>(`${API}/persons/${card.person_id}/name-suggestions`).subscribe({
      next:  s => this.nameSuggestions = s,
      // Diam-diam gagal — saran nama itu pemanis, bukan syarat mengisi form.
      error: () => {},
    });
  }

  pickSuggestion(s: NameSuggestion, event: Event): void {
    event.stopPropagation();
    this.nameInput    = s.name;
    this.jabatanInput = s.jabatan ?? '';
  }

  cancelNaming(event?: Event): void {
    event?.stopPropagation();
    this.namingPersonId = null;
    this.nameSuggestions = [];
  }

  confirmNaming(card: PersonCard, event: Event): void {
    event.stopPropagation();
    const name = this.nameInput.trim();
    if (!name) { this.cancelNaming(); return; }

    this.http.patch(`${API}/persons/${card.person_id}`, {
      name,
      jabatan: this.jabatanInput.trim() || null,
    }).subscribe({
      next: () => {
        this.namingPersonId = null;
        this._loadFeed();
        this._loadPersons();
      },
      error: err => { console.error('[people] beri nama error:', err); this.cancelNaming(); },
    });
  }

  // ── Helpers ──────────────────────────────────────────────────────────────────

  cameraName(camera_id: string): string {
    return this.cameras.find(c => c.camera_id === camera_id)?.name ?? camera_id;
  }

  cameraLocation(camera_id: string): string {
    return this.cameras.find(c => c.camera_id === camera_id)?.location || this.cameraName(camera_id);
  }

  personBadge(card: PersonCard): string {
    return `P-${card.person_id}`;
  }

  private _todayISO(): string {
    const d = new Date();
    return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}-${String(d.getDate()).padStart(2, '0')}`;
  }

  fmtTime2(iso: string | null): string {
    if (!iso) return '—';
    return new Date(iso).toLocaleTimeString('id-ID', { hour: '2-digit', minute: '2-digit' });
  }

  fmtDateTime(iso: string | null): string {
    if (!iso) return '—';
    return new Date(iso).toLocaleString('id-ID', { dateStyle: 'short', timeStyle: 'short' });
  }
}

File: backend/app/routers/people.py
from datetime import date

from fastapi import APIRouter, Query, Request

from app.schemas import PeopleFeedItem, PeopleFeedResponse, PersonsFeedResponse

router = APIRouter(prefix="/people", tags=["people"])


ATTR_THRESHOLD = 0.5  # skor mentah tracklets.attrs >= ini dianggap "punya atribut"
# Nama kunci di tracklets.attrs = string asli RAP1_ATTR_WORDS (lihat
# ai-service/app/par/par_service.py SELECTED_ATTRS), bukan nama ramah.
GENDER_ATTR = "female"


async def _build_filters(
    pool,
    q:           str | None,
    camera_id:   str | None,
    zone_id:     int | None,
    person_id:   int | None,
    from_date:   date | None,
    to_date:     date | None,
    upper_color: str | None,
    lower_color: str | None,
    gender:      str | None,
    attrs:       list[str] | None,
    similar_to:  int | None,
) -> tuple[list[str], list, str, str, "str | None"]:
    """Bangun WHERE/params/join yang dipakai bareng oleh /feed (per-deteksi) dan
    /persons (per-orang) — kedua endpoint menerima filter yang persis sama."""
    conditions: list[str] = []
    params: list = []
    needs_tracklet_join = any(v is not None for v in (upper_color, lower_color, gender)) \
        or bool(attrs) or similar_to is not None

    ref_embedding = None
    if similar_to is not None:
        ref_embedding = await pool.fetchval(
            "SELECT embedding::text FROM tracklets WHERE id = $1", similar_to
        )

    if camera_id:
        values = [v.strip() for v in camera_id.split(",") if v.strip()]
        if values:
            params.append(values)
            conditions.append(f"d.camera_id = ANY(${len(params)}::text[])")
    if zone_id is not None:
        params.append(zone_id)
        conditions.append(f"""EXISTS (
            SELECT 1 FROM zone_cameras zc2
            JOIN cameras c2 ON c2.id = zc2.camera_id
            WHERE c2.camera_id = d.camera_id AND zc2.zone_id = ${len(params)}
        )""")
    if person_id is not None:
        params.append(person_id)
        conditions.append(f"d.person_id = ${len(params)}")
    if q:
        params.append(f"%{q}%")
        conditions.append(f"(p.name ILIKE ${len(params)} OR p.label ILIKE ${len(params)})")
    if from_date:
        params.append(from_date)
        conditions.append(f"(d.timestamp AT TIME ZONE 'Asia/Jakarta')::date >= ${len(params)}")
    if to_date:
        params.append(to_date)
        conditions.append(f"(d.timestamp AT TIME ZONE 'Asia/Jakarta')::date <= ${len(params)}")
    if upper_color:
        values = [v.strip() for v in upper_color.split(",") if v.strip()]
        if values:
            params.append(values)
            conditions.append(f"t.attrs->>'upper_color' = ANY(${len(params)}::text[])")
    if lower_color:
        values = [v.strip() for v in lower_color.split(",") if v.strip()]
        if values:
            params.append(values)
            conditions.append(f"t.attrs->>'lower_color' = ANY(${len(params)}::text[])")
    if gender:
        op = ">=" if gender == "female" else "<"
        conditions.append(f"COALESCE((t.attrs->>'{GENDER_ATTR}')::float, 0) {op} {ATTR_THRESHOLD}")
    if attrs:
        for group in attrs:
            names = [n.strip() for n in group.split(",") if n.strip()]
            if not names:
                continue
            ors = []
            for name in names:
                params.append(name)
                ors.append(f"COALESCE((t.attrs->>${len(params)})::float, 0) >= {ATTR_THRESHOLD}")
            conditions.append("(" + " OR ".join(ors) + ")")
    if similar_to is not None:
        params.append(similar_to)
        conditions.append(f"d.tracklet_id != ${len(params)}")

    where         = f"WHERE {' AND '.join(conditions)}" if conditions else ""
    join_kind     = "JOIN" if similar_to is not None else "LEFT JOIN"  # similar_to butuh embedding, tidak boleh NULL
    tracklet_join = f"{join_kind} tracklets t ON t.id = d.tracklet_id" if needs_tracklet_join else ""

    return conditions, params, where, tracklet_join, ref_embedding


@router.get("/feed", response_model=PeopleFeedResponse)
async def people_feed(
    request:     Request,
    q:           str | None  = Query(None, description="Cari nama/label orang pemilik deteksi"),
    camera_id:   str | None  = Query(None, description="Boleh multi dipisah koma (OR) — 'c1,c2'"),
    zone_id:     int | None  = Query(None, description="Filter ke kamera yang tertaut ke zone ini"),
    person_id:   int | None  = Query(None, description="Semua deteksi satu orang — dipakai tab Timeline"),
    from_date:   date | None = Query(None, alias="from"),
    to_date:     date | None = Query(None, alias="to"),
    upper_color: str | None  = Query(None, description="Warna atasan, boleh multi dipisah koma (OR) — 'red,blue'"),
    lower_color: str | None  = Query(None, description="Warna bawahan, boleh multi dipisah koma (OR)"),
    gender:      str | None  = Query(None, description="'male' atau 'female', dari skor atribut PAR 'female'"),
    attrs:       list[str] | None = Query(None, description="Grup atribut PAR generik (mis. aksesoris: tas/topi/kacamata). Tiap value dipisah koma untuk OR di dalam grup; beberapa `attrs=` di-AND satu sama lain"),
    similar_to:  int | None  = Query(None, description="tracklet_id acuan — urutkan berdasar kemiripan penampilan (KNN pgvector)"),
    page:        int         = Query(1, ge=1),
    limit:       int         = Query(12, ge=1, le=100),
) -> PeopleFeedResponse:
    """Feed deteksi untuk halaman People. Tanpa filter = recent detections.
    `q`/`person_id` mempersempit ke satu orang. Filter atribut
    (upper_color/lower_color/gender/attrs) butuh tracklets.attrs terisi
    (hasil PAR) — deteksi yang belum punya tracklet tertaut otomatis tidak
    lolos filter atribut apa pun (LEFT JOIN, attrs NULL). `attrs` generik
    untuk grup PAR lain (aksesoris dll) — OR di dalam satu `attrs=`, AND
    antar `attrs=` berbeda. `similar_to` mengubah urutan hasil dari
    terbaru → terdekat embedding-nya (appearance-by-example) — filter lain
    tetap berlaku bersamaan."""
    pool = request.app.state.pool

    conditions, params, where, tracklet_join, ref_embedding = await _build_filters(
        pool, q, camera_id, zone_id, person_id, from_date, to_date,
        upper_color, lower_color, gender, attrs, similar_to,
    )
    if similar_to is not None and ref_embedding is None:
        return PeopleFeedResponse(items=[], total=0, page=page, pages=1, limit=limit)

    offset = (page - 1) * limit

    # COUNT tidak butuh urutan, jadi dijalankan dengan params SEBELUM
    # ref_embedding ditambahkan — kalau tidak, jumlah parameter yang dikirim
    # tidak cocok dengan placeholder $N yang benar-benar dipakai di SQL COUNT.
    total = await pool.fetchval(
        f"""
        SELECT COUNT(*)
        FROM detections d
        JOIN persons p ON p.id = d.person_id
        {tracklet_join}
        {where}
        """,
        *params,
    )

    if similar_to is not None:
        params = [*params, ref_embedding]
        order_clause = f"t.embedding <=> ${len(params)}::vector"
    else:
        order_clause = "d.timestamp DESC"

    rows = await pool.fetch(
        f"""
        SELECT
            d.id            AS detection_id,
            d.person_id     AS person_id,
            p.label         AS person_label,
            p.name          AS person_name,
            p.is_known      AS is_known,
            d.camera_id     AS camera_id,
            c.name          AS camera_name,
            c.location      AS camera_location,
            d.timestamp     AS timestamp,
            d.thumbnail_url AS thumbnail_url,
            d.tracklet_id   AS tracklet_id
        FROM detections d
        JOIN persons p       ON p.id = d.person_id
        LEFT JOIN cameras c  ON c.camera_id = d.camera_id
        {tracklet_join}
        {where}
        ORDER BY {order_clause}
        LIMIT {limit} OFFSET {offset}
        """,
        *params,
    )

    items = [PeopleFeedItem(**dict(r)) for r in rows]
    total_int = int(total or 0)
    return PeopleFeedResponse(
        items=items,
        total=total_int,
        page=page,
        pages=max(1, (total_int + limit - 1) // limit),
        limit=limit,
    )


@router.get("/persons", response_model=PersonsFeedResponse)
async def people_persons(
    request:     Request,
    q:           str | None  = Query(None, description="Cari nama/label orang"),
    camera_id:   str | None  = Query(None, description="Boleh multi dipisah koma (OR) — 'c1,c2'"),
    zone_id:     int | None  = Query(None, description="Filter ke kamera yang tertaut ke zone ini"),
    from_date:   date | None = Query(None, alias="from"),
    to_date:     date | None = Query(None, alias="to"),
    upper_color: str | None  = Query(None, description="Warna atasan, boleh multi dipisah koma (OR) — 'red,blue'"),
    lower_color: str | None  = Query(None, description="Warna bawahan, boleh multi dipisah koma (OR)"),
    gender:      str | None  = Query(None, description="'male' atau 'female', dari skor atribut PAR 'female'"),
    attrs:       list[str] | None = Query(None, description="Grup atribut PAR generik, sama seperti /feed"),
    similar_to:  int | None  = Query(None, description="tracklet_id acuan — tiap orang diranking pakai deteksi TERDEKATNYA ke acuan ini"),
    page:        int         = Query(1, ge=1),
    limit:       int         = Query(6, ge=1, le=100),
) -> PersonsFeedResponse:
    """Sama filternya dengan /feed, tapi paginasi per ORANG (bukan per
    deteksi) — satu baris per person_id, pakai deteksi terbarunya (atau
    terdekat ke `similar_to` kalau dipakai). Independen dari paginasi
    /feed: total/pages di sini dihitung dari jumlah distinct person, bukan
    jumlah deteksi, jadi frontend bisa punya dua kontrol halaman terpisah
    untuk section Orang dan Deteksi."""
    pool = request.app.state.pool

    conditions, params, where, tracklet_join, ref_embedding = await _build_filters(
        pool, q, camera_id, zone_id, None, from_date, to_date,
        upper_color, lower_color, gender, attrs, similar_to,
    )
    if similar_to is not None and ref_embedding is None:
        return PersonsFeedResponse(items=[], total=0, page=page, pages=1, limit=limit)

    offset = (page - 1) * limit

    total = await pool.fetchval(
        f"""
        SELECT COUNT(DISTINCT d.person_id)
        FROM detections d
        JOIN persons p ON p.id = d.person_id
        {tracklet_join}
        {where}
        """,
        *params,
    )

    if similar_to is not None:
        pick_order = f"t.embedding <=> ${len(params) + 1}::vector"
        params = [*params, ref_embedding]
    else:
        pick_order = "d.timestamp DESC"

    # Kalau similar_to: rank ORANG berdasar seberapa dekat deteksi TERBAIKNYA
    # (yang dipilih di CTE `latest`) ke embedding acuan — bukan waktu.
    outer_order = "l.dist ASC" if similar_to is not None else "l.timestamp DESC"
    dist_select = f", {pick_order} AS dist" if similar_to is not None else ", NULL::float AS dist"

    rows = await pool.fetch(
        f"""
        WITH latest AS (
            SELECT DISTINCT ON (d.person_id)
                d.id AS detection_id, d.person_id, d.camera_id, d.timestamp,
                d.thumbnail_url, d.tracklet_id
                {dist_select}
            FROM detections d
            JOIN persons p ON p.id = d.person_id
            {tracklet_join}
            {where}
            ORDER BY d.person_id, {pick_order}
        )
        SELECT
            l.detection_id  AS detection_id,
            l.person_id     AS person_id,
            p.label         AS person_label,
            p.name          AS person_name,
            p.is_known      AS is_known,
            l.camera_id     AS camera_id,
            c.name          AS camera_name,
            c.location      AS camera_location,
            l.timestamp     AS timestamp,
            l.thumbnail_url AS thumbnail_url,
            l.tracklet_id   AS tracklet_id
        FROM latest l
        JOIN persons p       ON p.id = l.person_id
        LEFT JOIN cameras c  ON c.camera_id = l.camera_id
        ORDER BY {outer_order}
        LIMIT {limit} OFFSET {offset}
        """,
        *params,
    )

    items = [PeopleFeedItem(**dict(r)) for r in rows]
    total_int = int(total or 0)
    return PersonsFeedResponse(
        items=items,
        total=total_int,
        page=page,
        pages=max(1, (total_int + limit - 1) // limit),
        limit=limit,
    )
